File: lib/widgets/speed_display.dart
import 'package:flutter/material.dart';

class SpeedDisplay extends StatefulWidget {
  final double speedValue;
  final double minSpeed;
  final double maxSpeed;

  const SpeedDisplay({
    super.key,
    required this.speedValue,
    required this.minSpeed,
    required this.maxSpeed,
  });

  @override
  _SpeedDisplayState createState() => _SpeedDisplayState();
}

class _SpeedDisplayState extends State<SpeedDisplay> {
  late double _speedValue;
  late double _minSpeed;
  late double _maxSpeed;

  @override
  void initState() {
    super.initState();
    _speedValue = widget.speedValue;
    _minSpeed = widget.minSpeed;
    _maxSpeed = widget.maxSpeed;
  }

  // Update local state when widget properties change.
  @override
  void didUpdateWidget(covariant SpeedDisplay oldWidget) {
    super.didUpdateWidget(oldWidget);
    bool needsUpdate = false;
    if (widget.speedValue != oldWidget.speedValue) {
      _speedValue = widget.speedValue;
      needsUpdate = true;
    }
    if (widget.minSpeed != oldWidget.minSpeed) {
      _minSpeed = widget.minSpeed;
      needsUpdate = true;
    }
    if (widget.maxSpeed != oldWidget.maxSpeed) {
      _maxSpeed = widget.maxSpeed;
      needsUpdate = true;
    }
    if (needsUpdate) {
      setState(() {}); // trigger rebuild if any value changed
    }
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final valueTextStyle = theme.textTheme.headlineLarge?.copyWith(
          fontSize: 96,
          fontWeight: FontWeight.bold,
        ) ??
        const TextStyle(fontSize: 96, fontWeight: FontWeight.bold);
    final minMaxValueTextStyle = theme.textTheme.headlineLarge?.copyWith(
          fontSize: 32,
          fontWeight: FontWeight.bold,
        ) ??
        const TextStyle(fontSize: 32, fontWeight: FontWeight.bold);
    final unitTextStyle = theme.textTheme.labelSmall?.copyWith(
          fontSize: 18,
        ) ??
        const TextStyle(fontSize: 18);
    final labelTextStyle = theme.textTheme.bodyMedium;

    return Column(
      children: [
        // Row with Min and Max speeds
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            // Min speed display
            Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text('Min', style: labelTextStyle),
                RichText(
                  text: TextSpan(
                    children: <TextSpan>[
                      TextSpan(
                        text: _minSpeed.toStringAsFixed(1),
                        style: minMaxValueTextStyle,
                      ),
                      TextSpan(
                        text: ' km/h',
                        style: unitTextStyle,
                      ),
                    ],
                  ),
                ),
              ],
            ),
            // Max speed display
            Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text('Max', style: labelTextStyle),
                RichText(
                  text: TextSpan(
                    children: <TextSpan>[
                      TextSpan(
                        text: _maxSpeed.toStringAsFixed(1),
                        style: minMaxValueTextStyle,
                      ),
                      TextSpan(
                        text: ' km/h',
                        style: unitTextStyle,
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ],
        ),
        const SizedBox(height: 16),
        // Single Row for main Speed display
        Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('Speed', style: labelTextStyle),
            RichText(
              text: TextSpan(
                children: <TextSpan>[
                  TextSpan(
                    text: _speedValue.toStringAsFixed(1),
                    style: valueTextStyle,
                  ),
                  TextSpan(
                    text: ' km/h',
                    style: unitTextStyle,
                  ),
                ],
              ),
            ),
          ],
        ),
      ],
    );
  }
}

File: src/widgets/speed_display.rs
use egui::{RichText, Ui, FontId, FontFamily};

pub struct SpeedDisplay {
    pub speed_value: f32,
    pub min_speed: f32,
    pub max_speed: f32,
}

impl SpeedDisplay {
    pub fn new(speed_value: f32, min_speed: f32, max_speed: f32) -> Self {
        Self {
            speed_value,
            min_speed,
            max_speed,
        }
    }

    pub fn ui(&self, ui: &mut Ui) {
        ui.vertical(|ui| {
            // Min and Max row
            ui.horizontal(|ui| {
                ui.vertical(|ui| {
                    ui.label("Min");
                    ui.horizontal(|ui| {
                        ui.label(
                            RichText::new(format!("{:.1}", self.min_speed))
                                .font(FontId::new(32.0, FontFamily::Monospace))
                                .strong()
                        );
                        ui.label(RichText::new(" km/h").size(14.0));
                    });
                });

                ui.add_space(40.0);

                ui.vertical(|ui| {
                    ui.label("Max");
                    ui.horizontal(|ui| {
                        ui.label(
                            RichText::new(format!("{:.1}", self.max_speed))
                                .font(FontId::new(32.0, FontFamily::Monospace))
                                .strong()
                        );
                        ui.label(RichText::new(" km/h").size(14.0));
                    });
                });
            });

            ui.add_space(16.0);

            // Main Speed
            ui.vertical_centered(|ui| {
                ui.label("Speed");
            });
            use egui::text::{LayoutJob, TextFormat};
            let mut job = LayoutJob::default();
            job.append(
                &format!("{:.1}", self.speed_value),
                0.0,
                TextFormat {
                    font_id: FontId::new(96.0, FontFamily::Monospace),
                    color: ui.visuals().text_color(),
                    ..Default::default()
                },
            );
            job.append(
                " km/h",
                0.0,
                TextFormat {
                    font_id: FontId::new(18.0, FontFamily::Proportional),
                    color: ui.visuals().text_color(),
                    ..Default::default()
                },
            );
            ui.vertical_centered(|ui| {
                ui.label(job);
            });
        });
    }
}
